### PR TITLE
Provide received_at in appropriate uplink messages

### LINKED ISSUE #1439: Application upstream message `received_at` never set
#### Summary

AS never sets `received_at` on the `ApplicationUp` message.

#### Steps to Reproduce

1. Subscribe to AS messages, i.e. MQTT or webhooks
2. See any upstream message, for example uplink

#### What do you see now?

No `received_at` set

#### What do you want to see instead?

For uplink-triggered messages (uplink, join-accept, downlink ack/nack), the `received_at` set to the `UplinkMessage` received from NS. In other scenarios (downlink queued/sent/failed, location solved, package-triggered messages, etc), don't set the value.

#### Can you do this yourself and submit a Pull Request?

Can review

### PR DESCRIPTION
**Pull request by adriansmares**
#### Summary

Closes #1439

#### Changes

- Pass the `received_at` from the `UplinkMessage` to the `ApplicationUp` that is sent to the ApplicationServer.

#### Notes for Reviewers

The initial issue mentioned that the AS should fill in the timestamp, but it was actually the NS which didn't fill it.

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, database and configuration, according to the stability commitments in `README.md`.
- [x] Testing: The changes are covered with unit tests. The changes are tested manually as well.
- [x] Documentation: Relevant documentation is added or updated.
- [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -17,7 +17,11 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Fixed
 
+<<<<<<< HEAD
```
Comment (rvolosatovs) [inline CHANGELOG.md]: :eyes:

Comment (adriansmares) [reply CHANGELOG.md]: Forgot to save before commiting.

**Comment by adriansmares** [comment]
> Why would NS set `ReceivedAt` in the application uplink?
> NS receives `UplinkMessage` and sets `ReceivedAt`, when it's received
> AS receives `ApplicationUp` and sets `ReceivedAt`, when it's received

My understanding from https://github.com/TheThingsNetwork/lorawan-stack/issues/1439#issue-504707230 is that the timestamp should actually be the one from the NS (`the received_at set to the UplinkMessage received from NS`). @johanstokking can you clarify this ?

**Comment by rvolosatovs** [comment]
Please see https://github.com/TheThingsNetwork/lorawan-stack/issues/1439#issuecomment-555030465 in the issue you linked

File: pkg/networkserver/networkserver_flow_test.go [hunk 2]
```diff
@@ -413,6 +414,7 @@ func handleClassAOTAAEU868FlowTest1_0_2(ctx context.Context, conn *grpc.ClientCo
 					Key: &appSKey,
 				},
 				SessionKeyID: []byte("session-key-id"),
+				ReceivedAt:   asUp.GetJoinAccept().GetReceivedAt(),
```
Comment (adriansmares) [inline pkg/networkserver/networkserver_flow_test.go]: @rvolosatovs do you see a way of achieving this without simulating the wall clock ? (we could do that in this test as well, if you consider it wouldn't be too artificial).

File: CHANGELOG.md [hunk 3]
```diff
@@ -8,6 +8,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ## [Unreleased]
 
 ### Added
+- The Network Server now provides the timestamp at which it received `JoinAccept` and `ApplicationUplink` messages.
```
Comment (htdvisser) [inline CHANGELOG.md]: Empty line between h3 and ul

File: api/messages.proto [hunk 4]
```diff
@@ -166,6 +166,7 @@ message ApplicationUplink {
   google.protobuf.Struct decoded_payload = 5;
   repeated RxMetadata rx_metadata = 6 [(validate.rules).repeated.min_items = 1];
   TxSettings settings = 7 [(gogoproto.nullable) = false, (validate.rules).message.required = true];
+  google.protobuf.Timestamp received_at = 8 [(gogoproto.stdtime) = true, (gogoproto.nullable) = false];
```
Comment (htdvisser) [inline api/messages.proto]: Maybe add a comment to explain _by_ what it is received at this time.

File: api/messages.proto [hunk 5]
```diff
@@ -184,6 +185,7 @@ message ApplicationJoinAccept {
   // Indicates whether the security context refers to the pending session, i.e. when this join-accept is an answer to a
   // rejoin-request.
   bool pending_session = 4;
+  google.protobuf.Timestamp received_at = 8 [(gogoproto.stdtime) = true, (gogoproto.nullable) = false];
```
Comment (htdvisser) [inline api/messages.proto]: Maybe add a comment to explain _by_ what it is received at this time.

File: pkg/networkserver/networkserver_flow_test.go (hunk 2 above)
Comment (rvolosatovs) [reply pkg/networkserver/networkserver_flow_test.go]: This looks good to me - I don't we should mock time in flow tests.

File: CHANGELOG.md (hunk 3 above)
Comment (rvolosatovs) [inline CHANGELOG.md]:
```suggestion
- The Network Server now provides the timestamp at which it received join-accept or data uplink messages
```

or

```suggestion
- The Network Server now provides the timestamp at which it received `JoinAccept` and `UplinkMessage` messages.
```

In any case:
a. NS does not receive `ApplicationUplink`
b. I don't think we should use proto message names here - we are writing this for users, they should need to look through our source code to understand what a particular change means

**Comment by htdvisser** [comment]
This PR has API changes that are visible externally, so it requires a minor version bump.

**Comment by johanstokking** [comment]
> This PR has API changes that are visible externally, so it requires a minor version bump.

I agree it requires a minor bump, but not for this reason. We should be safely _adding_ externally visible fields. The reason for a minor bump is that we're changing API between components (NS and AS) for this to work.

**Pull request merged**